### PR TITLE
Fix iteration in _remove_deleted_email_pushers background job.

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/10729, see issue for context.

We were originally iterating over `txn`, which can work, but as `txn` was being used to run a
simple_delete operation inside the loop, it was actually being redefined mid-loop.

Thus, this caused the background job that was intended to remove old pushers that were
associated with an unlinked email address to fail if there were any to remove.We were originally iterating over 'txn', which can work, but as 'txn' was being used to run a
simple_delete operation inside the loop, it was actually being redefined mid-loop.

Thus, this caused the background job that was intended to remove old pushers that were
associated with an unlinked email address to fail if there were any to remove.

### CODE REVIEW

File: tests/push/test_email.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def make_homeserver(self, reactor, clock):
         config["public_baseurl"] = "aaa"
         config["start_pushers"] = True
 
-        hs = self.setup_test_homeserver(config=config)
+        self.hs = self.setup_test_homeserver(config=config)
```
Comment (richvdh) [inline tests/push/test_email.py:70]: `setUp` (the thing that calls `make_homeserver` sets `self.hs`, so this is unnecessary.

File: tests/push/test_email.py [hunk 2]
```diff
@@ -344,6 +344,50 @@ def test_no_email_sent_after_removed(self):
...
+        self.hs.get_datastore().db_pool.updates._all_done = False
+
+        # Now let's actually drive the updates to completion
+        while not self.get_success(
+            self.hs.get_datastore().db_pool.updates.has_completed_background_updates()
+        ):
+            self.get_success(
+                self.hs.get_datastore().db_pool.updates.do_next_background_update(100),
+                by=0.1,
+            )
```
Comment (richvdh) [inline tests/push/test_email.py:382]: We should probably extract all this to a utility function. Maybe next time.

**Review by richvdh — COMMENTED** [review]
needs a newsfile. looks sensible otherwise.

File: tests/push/test_email.py (hunk 1 above)
Comment (anoadragon453) [reply tests/push/test_email.py]: Aha, indeed, good point!

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**